src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/SpecializedTasks.cs

```diff
@@ -86,6 +87,88 @@ public static ValueTask<T[]> WhenAll<T>(IEnumerable<ValueTask<T>> tasks)
+        /// <param name="arg">The state to pass to <paramref name="func"/> and <paramref name="transform"/>.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> that the operation will observe.</param>
+        /// <returns></returns>
+        public static ValueTask<TResult> TransformWithoutIntermediateCancellationExceptionAsync<TArg, TIntermediate, TResult>(
+            Func<TArg, CancellationToken, ValueTask<TIntermediate>> func,
+            Func<TIntermediate, TArg, TResult> transform,
+            TArg arg,
+            CancellationToken cancellationToken)
+        {
+            if (func is null)
+                throw new ArgumentNullException(nameof(func));
+            if (transform is null)
+                throw new ArgumentNullException(nameof(transform));
+
+            var intermediateResult = func(arg, cancellationToken);
+            if (intermediateResult.IsCompletedSuccessfully)
+            {
+                // Synchronous fast path if 'func' completes synchronously
+                var result = intermediateResult.Result;
+                if (cancellationToken.IsCancellationRequested)
+                    return new ValueTask<TResult>(Task.FromCanceled<TResult>(cancellationToken));
+
+                return new ValueTask<TResult>(transform(result, arg));
+            }
+            else if (intermediateResult.IsCanceled && cancellationToken.IsCancellationRequested)
+            {
+                // Synchronous fast path if 'func' cancels synchronously
+                return new ValueTask<TResult>(Task.FromCanceled<TResult>(cancellationToken));
+            }
+            else
+            {
+                // Asynchronous fallback path
+                return UnwrapAndTransformAsync(intermediateResult, transform, arg, cancellationToken);
+            }
+
+            static ValueTask<TResult> UnwrapAndTransformAsync(ValueTask<TIntermediate> intermediateResult, Func<TIntermediate, TArg, TResult> transform, TArg arg, CancellationToken cancellationToken)
+            {
+                // Apply the transformation function once a result is available. The behavior depends on the final
+                // status of 'intermediateResult' and the 'cancellationToken'.
+                //
+                // | 'intermediateResult'       | 'cancellationToken' | Behavior                                 |
+                // | -------------------------- | ------------------- | ---------------------------------------- |
+                // | Ran to completion          | Not cancelled       | Apply transform                          |
+                // | Ran to completion          | Cancelled           | Cancel result without applying transform |
+                // | Cancelled (matching token) | Cancelled           | Cancel result without applying transform |
+                // | Cancelled (mismatch token) | Not cancelled       | Cancel result without applying transform |
```

Comment:
this one is the most interesting.  woudl that indicate a bug?

Reply:
➡️ This is _almost_ the same behavior you have for normal `async`/`await`. The only difference is this helper will produce a result task with the input cancellation token to the method instead of using the one included in the nested cancellation.